framework/src/utils/RankTwoTensor.C

```diff
@@ -889,6 +927,263 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+{
+  mooseAssert(N == old_row.size() && N == new_row.size(),
+              "rowPermutationTensor: number of rows should be equal to LIBMESH_DIM");
+
+  RankTwoTensorTempl<T> P;
+
+  for (unsigned int i = 0; i < N; ++i)
+    P(old_row[i], new_row[i]) = 1.0;
+
+  return P;
+}
+
+template <typename T>
+RankTwoTensorTempl<T>
+RankTwoTensorTempl<T>::columnPermutationTensor(std::vector<unsigned int> old_col,
+                                               std::vector<unsigned int> new_col) const
+{
+  RankTwoTensorTempl<T> P = this->rowPermutationTensor(old_col, new_col);
+  return P.transpose();
+}
+
+template <typename T>
+RankTwoTensorTempl<T>
+RankTwoTensorTempl<T>::givensRotation(unsigned int row1, unsigned int row2, unsigned int col) const
+{
+  T c, s;
+  T a = (*this)(row1, col);
+  T b = (*this)(row2, col);
+
+  if (MooseUtils::absoluteFuzzyEqual(b, 0.0))
+  {
+    c = a >= 0.0 ? 1.0 : -1.0;
+    s = 0.0;
+  }
+  else if (MooseUtils::absoluteFuzzyEqual(a, 0.0))
+  {
+    c = 0.0;
+    s = b >= 0.0 ? 1.0 : -1.0;
+  }
+  else if (std::abs(a) > std::abs(b))
+  {
+    T t = b / a;
+    Real sgn = a >= 0.0 ? 1.0 : -1.0;
+    T u = sgn * std::sqrt(1.0 + t * t);
+    c = 1.0 / u;
+    s = c * t;
+  }
+  else
+  {
+    T t = a / b;
+    Real sgn = b >= 0.0 ? 1.0 : -1.0;
+    T u = sgn * std::sqrt(1.0 + t * t);
+    s = 1.0 / u;
+    c = s * t;
+  }
+
+  RankTwoTensorTempl<T> R(initIdentity);
+  R(row1, row1) = c;
+  R(row1, row2) = s;
+  R(row2, row1) = -s;
+  R(row2, row2) = c;
+
+  return R;
+}
+
+template <>
+RankTwoTensorTempl<DualReal>
+RankTwoTensorTempl<DualReal>::givensRotation(unsigned int row1,
+                                             unsigned int row2,
+                                             unsigned int col) const
+{
+
+  DualReal cos = 0.0;
+  DualReal sin = 0.0;
+  DualReal a = (*this)(row1, col);
+  DualReal b = (*this)(row2, col);
+  RankTwoTensorTempl<DualReal> R(initIdentity);
+
+  // rotate Real number
+  RankTwoTensorTempl<Real> temp;
+  temp(row1, col) = a.value();
+  temp(row2, col) = b.value();
+  temp = temp.givensRotation(row1, row2, col);
+  cos.value() = temp(row1, row1);
+  sin.value() = temp(row1, row2);
+
+  // rotate Dual number
+  for (unsigned int i = 0; i < AD_MAX_DOFS_PER_ELEM; i++)
+  {
+    if (MooseUtils::absoluteFuzzyEqual(a.value(), 0.0) &&
+        MooseUtils::absoluteFuzzyEqual(b.value(), 0.0))
+    {
+      cos.derivatives()[i] = 0.0;
+      sin.derivatives()[i] = 0.0;
+    }
+    else
+    {
+      cos.derivatives()[i] = (a.derivatives()[i] * sin.value() * sin.value() -
+                              b.derivatives()[i] * cos.value() * sin.value()) /
+                             (b.value() * sin.value() + a.value() * cos.value());
+      sin.derivatives()[i] = (b.derivatives()[i] * cos.value() * cos.value() -
+                              a.derivatives()[i] * cos.value() * sin.value()) /
+                             (b.value() * sin.value() + a.value() * cos.value());
```

Comment:
why do you need to manually update the derivatives here? Why does the automatic differentiation not work?

Reply:
I wrote this before I fixed the QR algorithm. We shouldn't need this anymore. Let me remove it.

The very issue of this entire algorithm is that when both entries to rotate are 0. There is no way to rotate the dual numbers (imagine rotating a zero-length vector), which is why I did additional permutation before doing the rotation.